SpatialGDK/Source/SpatialGDK/Public/Interop/RPCs/RPCTypes.h

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "Interop/Connection/SpatialGDKSpanId.h"
+#include "Schema/RPCPayload.h"
+#include "SpatialView/EntityView.h"
+
+namespace SpatialGDK
+{
+namespace RPCCallbacks
+{
+using DataWritten = TFunction<void(Worker_EntityId, Worker_ComponentId, Schema_ComponentData*)>;
+using UpdateWritten = TFunction<void(Worker_EntityId, Worker_ComponentId, Schema_ComponentUpdate*)>;
+using RequestWritten = TFunction<void(Worker_EntityId, Schema_CommandRequest*)>;
+using ResponseWritten = TFunction<void(Worker_EntityId, Schema_CommandResponse*)>;
+using RPCWritten = TFunction<void(Worker_EntityId, Worker_ComponentId, uint32)>;
+
+using CanExtractRPCs = TFunction<bool(Worker_EntityId)>;
+using ProcessRPC = TFunction<bool(Worker_EntityId, const RPCPayload&)>;
+} // namespace RPCCallbacks
+/**
+ * Structure encapsulating a read operation
+ */
+struct RPCReadingContext
+{
+	Worker_EntityId EntityId;
+	Worker_ComponentId ComponentId;
+
+	bool IsUpdate() const { return Update != nullptr; }
+
+	Schema_ComponentUpdate* Update = nullptr;
+	Schema_Object* Fields = nullptr;
+};
+
+/**
+ * Structure encapsulating a write operation.
+ * It serves as a factory for EntityWrites which encapsulate writes to a given entity/component pair
+ */
+struct RPCWritingContext
+{
+	enum class DataKind
+	{
+		Generic,
+		ComponentData,
+		ComponentUpdate,
+		CommandRequest,
+		CommandResponse
+	};
+
+	/**
+	 * RAII object to encapsulate writes to an entity/component couple.
+	 * It makes sure that the appropriate callback is executed when the write operation is done
+	 */
+	class EntityWrite
+	{
+	public:
+		EntityWrite(const EntityWrite&) = delete;
+		EntityWrite& operator=(const EntityWrite&) = delete;
+		EntityWrite& operator=(EntityWrite&&) = delete;
+
+		EntityWrite(EntityWrite&& Write);
+		~EntityWrite();
+
+		Schema_ComponentUpdate* GetComponentUpdateToWrite();
+		Schema_Object* GetFieldsToWrite();
+
+		// Must be called by writers to allow any per-RPC operation to take place.
+		// The RPCId parameter is intended to be unique in the EntityId/ComponentId context.
+		void RPCWritten(uint32 RPCId);
+
+		const Worker_EntityId EntityId;
+		const Worker_ComponentId ComponentId;
+
+	private:
+		union
+		{
+			Schema_GenericData* GenData;
+			Schema_ComponentData* Data;
+			Schema_ComponentUpdate* Update;
+			Schema_CommandRequest* Request;
+			Schema_CommandResponse* Response;
+		};
+
+		EntityWrite(RPCWritingContext& InCtx, Worker_EntityId InEntityId, Worker_ComponentId InComponentID);
```

Comment:
I am personally not a fan of having the constructor this low in the file, is there a reason for this?

Reply:
In this case, because it is private.

Reply:
Would prefer moving it up with a private block, especially because the intricate nature of this class I think it would help with clarity to follow the convention of defining the constructor on top.

Reply:
I have a different POV on that matter, I keep the public API at the top, because it's what you are interested with first as a user.
If you are the maintainer/implementer, then you have to consider the entire class anyway.

Reply:
Ok a minor issue anyway I guess.